WalletWasabi/WabiSabi/Client/CredentialDependencies/DependencyGraph.cs

```diff
@@ -0,0 +1,350 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+
+namespace WalletWasabi.WabiSabi.Client.CredentialDependencies
+{
+	public class DependencyGraph
+	{
+		public const int K = ProtocolConstants.CredentialNumber;
+
+		private DependencyGraph(IEnumerable<IEnumerable<long>> initialValues)
+		{
+			var initialValuesImmutable = initialValues.Select(x => x.ToImmutableArray()).ToImmutableArray();
+			if (initialValuesImmutable.Any(x => x.Length != (int)CredentialType.NumTypes))
+			{
+				throw new ArgumentException($"Number of credential values must be {CredentialType.NumTypes}");
+			}
+			if (initialValuesImmutable.Any(x => x.Where(y => y != 0).Select(y => y.CompareTo(0)).Distinct().Count() != 1))
+			{
+				throw new ArgumentException($"Credential values of a node must all have the same sign.");
+			}
+
+			for (CredentialType i = 0; i < CredentialType.NumTypes; i++)
+			{
+				if (initialValuesImmutable.Sum(x => x[(int)i]) < 0)
+				{
+					throw new ArgumentException("Overall balance must not be negative");
+				}
+			}
+
+			// per node entries created in AddNode, querying nodes not in the
+			// graph should result in key errors.
+
+			Successors = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, HashSet<CredentialDependency>>()).ToImmutableArray();
+			Predecessors = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, HashSet<CredentialDependency>>()).ToImmutableArray();
+			EdgeBalances = Enumerable.Repeat(0, (int)CredentialType.NumTypes).Select(_ => new Dictionary<int, long>()).ToImmutableArray();
+
+			Vertices = ImmutableList<RequestNode>.Empty;
+
+			foreach (var node in initialValuesImmutable.Select((values, i) => new RequestNode(i, values)))
+			{
+				// enforce at least one value != 0? all values? it doesn't actually matter for the algorithm
+				AddNode(node);
+			}
+		}
+
+		public ImmutableList<RequestNode> Vertices { get; private set; }
+
+		// Internal properties used to keep track of effective values and edges
+		private ImmutableArray<Dictionary<int, long>> EdgeBalances { get; }
+		private ImmutableArray<Dictionary<int, HashSet<CredentialDependency>>> Successors { get; }
+		private ImmutableArray<Dictionary<int, HashSet<CredentialDependency>>> Predecessors { get; }
+
+		// TODO doc comment
+		// Public API: construct a graph from amounts, and resolve the
+		// credential dependencies. Should only produce valid graphs.
+		// IDs are positive ints assigned in the order of the enumerable, but
```

Comment:
What IDs?

Reply:
Since there are potentially more nodes in the resulting graph than the number of input/output amounts given to the constructor, every reissuance request needs a node, numerical IDs are allocated for them.

These IDs are numerical ascending for the input values arguments, so they correspond with the indexes given to `ResolveCredentialDependencies(amounts)`, so you can use `graph.Vertices[i]` to interrogate the graph for the edges for `amount[i]`.

I'm actually thinking of making the `ID` property private or removing it entirely, so that the only way to map between them is the ordering, as I think that's enough and the index is implicit from the order of `Vertices`, in which case the internal dictionaries won't need an `int` as the key anymore, they can just use the `RequestNode` directly (but it needs to not be a record type for that).